packages/next/server/render.tsx

```diff
@@ -1467,49 +1537,107 @@ function renderToNodeStream(
   })
 }
 
-function renderToReadableStream(
+function connectReactServerReadableStreamToPiper(
+  write: (s: string) => void,
+  next: (err?: Error) => void
+) {
+  let bufferedString = ''
+  let flushTimeout: null | NodeJS.Timeout = null
+
+  function flushBuffer() {
+    // Intentionally delayed writing when using ReadableStream due to the lack
+    // of cork/uncork APIs.
+    if (!flushTimeout) {
+      flushTimeout = setTimeout(() => {
+        write(bufferedString)
+        bufferedString = ''
+        flushTimeout = null
+      }, 0)
+    }
+  }
+
+  function startWriting(reader: ReadableStreamDefaultReader) {
+    const decoder = new TextDecoder()
+    const process = () => {
+      reader.read().then(({ done, value }: any) => {
+        if (!done) {
+          const s = typeof value === 'string' ? value : decoder.decode(value)
+          bufferedString += s
+          flushBuffer()
+          process()
+        } else {
+          // Make sure it's scheduled after the current flushing.
+          setTimeout(() => next(), 0)
+        }
+      })
+    }
+    process()
+  }
+
+  return {
+    startWriting,
+  }
+}
+
+function renderToWebStream(
```

Comment:
Not super sure what's happening here. You don't need (or really want, since it could lead to early buffering) a `startWriting` -- you should just need to start reading from the stream inside the returned piper function.

I'd just move the buffering up to the middleware layer.

Reply:
Something was introduced in #31186, to delay resolving until `onCompleteShell` happens. If there is an error occurred before that, we need to throw, otherwise we can safely return the piper.